crates/lib/src/action/worker.rs

```diff
@@ -32,17 +32,140 @@ pub async fn worker(
 	events_tx: priority::Sender<Event, Priority>,
 	events: priority::Receiver<Event, Priority>,
 ) -> Result<(), CriticalError> {
-	let mut last = Instant::now();
-	let mut set = Vec::new();
-	let process = ProcessHolder::default();
-	let outcome_gen = OutcomeWorker::newgen();
+	let mut processes: HashMap<SupervisorId, ProcessData> = HashMap::new();
+
+	// Insert the base process from the initial `WorkingData`.
+	let base_process_id = SupervisorId::default();
+	processes.insert(
+		base_process_id,
+		ProcessData {
+			working: working.clone(),
+			commands: working.borrow().commands.clone(),
+			process: ProcessHolder::default(),
+			outcome_gen: OutcomeWorker::newgen(),
+		},
+	);
+
+	while let Some(mut set) = throttle_collect(
+		events.clone(),
+		working.clone(),
+		errors.clone(),
+		Instant::now(),
+	)
+	.await?
+	{
+		#[allow(clippy::iter_with_drain)]
+		let events = Arc::from(set.drain(..).collect::<Vec<_>>().into_boxed_slice());
+		let action = Action::new(
+			Arc::clone(&events),
+			Arc::from(
+				processes
+					.keys()
+					.copied()
+					.collect::<Vec<_>>()
+					.into_boxed_slice(),
+			),
+		);
+		debug!("running action handler");
+		let action_handler = {
+			let wrk = working.borrow();
+			wrk.action_handler.clone()
+		};
 
-	loop {
-		if events.is_closed() {
-			trace!("events channel closed, stopping");
-			break;
+		let outcomes = action.outcomes.clone();
+		let err = action_handler
+			.call(action)
+			.await
+			.map_err(|e| rte("action worker", e.as_ref()));
+		if let Err(err) = err {
+			errors.send(err).await?;
+			debug!("action handler errored, skipping");
+			continue;
 		}
 
+		for (pid, (resolution, event_set)) in outcomes.lock().await.iter() {
+			let found = match resolution {
```

Comment:
would appreciate debug! logs for each branch here

Reply:
would something like this be the format you are looking for in the `debug!`'s?

```rust
debug!(pid=?pid, cmds=?cmds, "starting new command");
```